drivers/tipset_message_producer.go

```diff
@@ -11,125 +12,151 @@ import (
 type TipSetMessageBuilder struct {
 	driver *TestDriver
 
+	bbs []*BlockBuilder
+}
+
+func NewTipSetMessageBuilder(testDriver *TestDriver) *TipSetMessageBuilder {
+	return &TipSetMessageBuilder{
+		driver: testDriver,
+		bbs:    nil,
+	}
+}
+
+func (t *TipSetMessageBuilder) WithBlockBuilder(bb *BlockBuilder) *TipSetMessageBuilder {
+	t.bbs = append(t.bbs, bb)
+	return t
+}
+
+func (t *TipSetMessageBuilder) Apply() []types.MessageReceipt {
+	var blks []types.BlockMessagesInfo
+	for _, b := range t.bbs {
+		blks = append(blks, b.build())
+	}
+	receipts, err := t.driver.validator.ApplyTipSetMessages(t.driver.ExeCtx.Epoch, t.driver.State(), blks, t.driver.Randomness())
+	require.NoError(t.driver.T, err)
+
+	return receipts
+}
+
+func (t *TipSetMessageBuilder) ApplyAndValidate() {
+	receipts := t.Apply()
+
+	var results []Result
+	for _, b := range t.bbs {
+		results = append(results, b.expectedResults...)
+	}
+
+	if len(receipts) > len(results) {
+		t.driver.T.Fatalf("ApplyTipSetMessages returned more receipts than expected. Expected: %d, Actual: %d", len(results), len(receipts))
+	}
+
+	for i := range receipts {
```

Comment:
It would be a little more helpful to iterate over the blocks, rather than flatten, so that the `(blockIndex, messageIndex)` (before any de-duplication) is available.

Reply:
Another way of saying this: push validation into the BlockBuilder, so that here we iterate over blocks calling ApplyAndValidate, and thus can know the block index and message index in the block. Duplicate messages makes it a little tricky. you might need an explicit null result placeholder for the dropped messages.

Reply:
Since this is non-urgent (a possibly holding up other work to avoid rebase-hell) I will add this in a follow on. Nice idea!